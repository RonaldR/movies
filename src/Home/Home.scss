body {
	background: white;

	@media (prefers-color-scheme: dark) {
		background: #111111;
	}
}

.App {
	width: 100%;
	max-width: 640px;
	margin: 0 auto;
}

form {
	width: 100%;
}

.search {
	position: relative;

	&::after {
		content: 'ðŸŽ¬';
		width: 32px;
		height: 32px;
		position: absolute;
		right: 0;
		top: 10px;
		z-index: 1;
		font-size: 20px;
		opacity: 0;
		animation-duration: 2s;
		animation-iteration-count: infinite;
		transform-origin: bottom;
		animation-name: bounce;
		animation-timing-function: cubic-bezier(0.28, 0.84, 0.42, 1);
	}

	&.loading::after {
		opacity: 1;
	}
}

@keyframes bounce {
	0% {
		transform: scale(1, 1) translateY(0);
	}
	10% {
		transform: scale(1.1, 0.9) translateY(0);
	}
	30% {
		transform: scale(0.9, 1.1) translateY(-10px);
	}
	50% {
		transform: scale(1.05, 0.95) translateY(0);
	}
	57% {
		transform: scale(1, 1) translateY(-3px);
	}
	64% {
		transform: scale(1, 1) translateY(0);
	}
	100% {
		transform: scale(1, 1) translateY(0);
	}
}

.search__input {
	padding: 12px;
	width: 100%;
	background: #fafafa;
	border: 1px solid #e8e8e8;
	font-size: 24px;
	outline: none;
}

.movies {
	list-style: none;
	display: flex;
	flex-direction: column;
	align-items: center;
	width: 100%;
	max-width: 640px;
	padding: 0;
	background: #fafafa;
	border: 1px solid #e8e8e8;
	border-top: 0;
	margin: 0;
}
