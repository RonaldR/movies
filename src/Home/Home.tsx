import React, { useState, useEffect } from 'react';
import { useDebouncedCallback } from 'use-debounce';

import MovieCard from '../MovieCard/MovieCard';
import { searchMovie } from '../utils/api';
import { MovieSearch } from '../types/types';

import './Home.scss';

const App = () => {
  const [movies, setMovies] = useState<Array<MovieSearch> | undefined>();
  const [loading, setLoading] = useState(true);

  const getMovies = async (search: string = 'matrix') => {
    setLoading(true);
    const response = await searchMovie(search);
    const json = await response.json();
    setLoading(false);
    return json.Search;
  }

  useEffect(() => {
    // IIFE
    (async () => {
      setMovies(await getMovies())
    })();
  }, []);

  // Debounce callback
  const [debouncedCallback] = useDebouncedCallback(
    // function
    async (value: string) => {
      setMovies(await getMovies(value));
    },
    // delay in ms
    200
  );

  return (
    <main className="App">
      <form onSubmit={(e) => e.preventDefault()}>
        <div className={`search ${loading ? 'loading' : ''}`}>
          <input className="search__input" type="text" placeholder="Search movie (Matrix)" onChange={(e) => debouncedCallback(e.target.value)} data-testid="search" />
        </div>
      </form>

      <ul className="movies">
        {movies && movies.map(movie => (
          <MovieCard key={movie.imdbID} movie={movie} />
        ))}
      </ul>
    </main>
  );
}

export default App;
