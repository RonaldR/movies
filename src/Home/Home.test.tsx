import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import Home from './Home';

// TODO: add more tests and api mock

describe('Home', () => {
	test('renders search input', () => {
		const { getByTestId, asFragment } = render(<Home />);
		const searchInput = getByTestId('search');
		expect(searchInput).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});

	test('search input value can change', () => {
		const { getByTestId } = render(<MemoryRouter><Home /></MemoryRouter>);
		const searchInput = getByTestId('search');
		fireEvent.change(searchInput, { target: { value: '23' } });
		expect(searchInput.value).toBe('23');
	});

	test('api call function', async () => {
		const { getByTestId, asFragment } = render(<MemoryRouter><Home /></MemoryRouter>);
		const searchInput = getByTestId('search');
		fireEvent.change(searchInput, { target: { value: '23' } });
		expect(asFragment()).toMatchSnapshot();
	});
});
