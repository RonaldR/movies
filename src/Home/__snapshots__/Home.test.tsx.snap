// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home api call function 1`] = `
<DocumentFragment>
  <main
    class="App"
  >
    <form>
      <div
        class="search loading"
      >
        <input
          class="search__input"
          data-testid="search"
          placeholder="Search movie (Matrix)"
          type="text"
        />
      </div>
    </form>
    <ul
      class="movies"
    />
  </main>
</DocumentFragment>
`;

exports[`Home renders search input 1`] = `
<DocumentFragment>
  <main
    class="App"
  >
    <form>
      <div
        class="search loading"
      >
        <input
          class="search__input"
          data-testid="search"
          placeholder="Search movie (Matrix)"
          type="text"
        />
      </div>
    </form>
    <ul
      class="movies"
    />
  </main>
</DocumentFragment>
`;
